Beautification of Examples (#491)
zenml-io/zenml: docs/book/developer-guide/materializer.md

`StatisticsArtifact`, `DriftArtifact`, etc. This is simply a tag to query certain artifact types in the post-execution 
workflow.

### Extending the `BaseMaterializer`

> Let's say you a custom object called `MyObject` that flows between two steps in a pipeline:

Reviewer: ```suggestion
Let's say you have a custom class called `MyObject` that flows between two steps in a pipeline:
```

```

The above basically means that ZenML does not know how to persist the object of type `MyObj` between steps (how could 

> it? We just created this!). Therefore, we can have to create our own materializer. To do this you can simply extend the 

Reviewer: I can't comment on the error message above, but that contains a wrong URL to our docs. We will need to fix this both in code and here to point to the updated version.

Author: Definitely, catching outdated urls is gonna be quite the task

> it? We just created this!). Therefore, we can have to create our own materializer. To do this you can simply extend the 

Reviewer: ```suggestion
it? We just created this!). Therefore, we have to create our own materializer. To do this you can simply extend the 
```

> Pro-tip: Use the ZenML `fileio` handler to ensure your materialization logic works across artifact stores (local and 

Reviewer: ```suggestion
Pro-tip: Use the ZenML `fileio` module to ensure your materialization logic works across artifact stores (local and 
```

Should we maybe mark this as a hint/warning?

remote like S3 buckets).

Now ZenML can use this materializer to handle outputs and inputs of your customs object. Edit the pipeline as follows 
to see this in action:

```python
first_pipeline(

>     step_1=my_first_step().with_return_materializers(MyMaterializer),

Reviewer: I'm not 100% sure, but how I understand it you wouldn't even need to specify the `with_return_materializer` right?  As the class `MyObj` is already an associated type of the materializer, it should automatically detect it. Not sure if we should mention that functionality somehow

Author: Very good question, I'll try to find a way of putting a sentence in there about this. I feel Materializers are the most misunderstood part of ZenML atm, so it makes sense to put some elbow grease in here

```

### Summary in Code

<details>
    <summary>Code Example of this Section</summary>

```python
import os
from typing import Type
import logging

from zenml.steps import step
from zenml.pipelines import pipeline

from zenml.artifacts import DataArtifact
from zenml.io import fileio
from zenml.materializers.base_materializer import BaseMaterializer

class MyObj:
    def __init__(self, name: str):
        self.name = name

class MyMaterializer(BaseMaterializer):
    ASSOCIATED_TYPES = (MyObj,)
    ASSOCIATED_ARTIFACT_TYPES = (DataArtifact,)

    def handle_input(self, data_type: Type[MyObj]) -> MyObj:
        """Read from artifact store"""
        super().handle_input(data_type)
        with fileio.open(os.path.join(self.artifact.uri, 'data.txt'),
                         'r') as f:
            name = f.read()
        return MyObj(name=name)

    def handle_return(self, my_obj: MyObj) -> None:
        """Write to artifact store"""
        super().handle_return(my_obj)
        with fileio.open(os.path.join(self.artifact.uri, 'data.txt'),
                         'w') as f:
            f.write(my_obj.name)

> print

Reviewer: ```suggestion

```
